src/skimage/registration/_affine.py

```diff
@@ -0,0 +1,549 @@
+        Intial guess for the homogeneous transformation matrix.
+    model : {'affine', 'euclidean', 'translation'}
+        Motion model: "translation", "euclidean" or "affine" where euclidean motion
+        corresponds to rigid motion (rotation and translation) and affine motion can
+        represent change in scale, shear, rotation and translations.
+    solver: lambda(reference_image, moving_image, weights, channel_axis, matrix) -> matrix
+        Affine motion solver can be solver_affine_lucas_kanade or solver_affine_studholme.
+    pyramid_scale : float
+        The pyramid downscale factor.
+    pyramid_minimum_size : int
+        The minimum size for any dimension of the pyramid levels.
+
+    Returns
+    -------
+    matrix : ndarray
+        The transformation matrix.
+
+    Note
+    ----
+    Generate image pyramids and apply the solver at each level passing the
+    previous affine parameters from the previous estimate.
+
+    The estimated matrix can be used with func:`scipy.ndimage.affine` to register the moving image
+    to the reference image.
+
+    Reference
+    ---------
+    .. [1] J. Nunez-Iglesias, S. van der Walt, and H. Dashnow, Elegant SciPy:
+           The Art of Scientific Python. O’Reilly Media, Inc., 2017.
+
+    Example
+    -------
+    >>> from skimage import data
+    >>> import numpy as np
+    >>> from scipy import ndimage as ndi
+    >>> from skimage.registration import affine
+    >>> reference = data.camera()
+    >>> r = -0.12  # radians
+    >>> c, s = np.cos(r), np.sin(r)
+    >>> transform = np.array([[c, -s, 0], [s, c, 0], [0, 0, 1]])
+    >>> moving = ndi.affine_transform(reference, transform)
+    >>> matrix = affine(reference, moving)
+    >>> registered = ndi.affine_transform(moving, matrix)
+    """
+
+    ndim = reference_image.ndim if channel_axis is None else reference_image.ndim - 1
+
+    if channel_axis is not None:
+        shape = [d for k, d in enumerate(reference_image.shape) if k != channel_axis]
+    else:
+        shape = reference_image.shape
+        if min(shape) <= 6:
+            warnings.warn(f"No channel axis specified for shape {shape}")
+
+    max_layer = int(
+        floor(
+            log(min(shape), pyramid_downscale)
+            - log(pyramid_minimum_size, pyramid_downscale)
+        )
+    )
+
+    pyramid = list(
+        zip(
+            *[
+                reversed(
+                    list(
+                        pyramid_gaussian(
+                            x,
+                            max_layer=max_layer,
+                            downscale=pyramid_downscale,
+                            preserve_range=True,
+                            channel_axis=channel_axis,
+                        )
+                        if x is not None
+                        else [None] * (max_layer + 1)
+                    )
+                )
+                for x in [reference_image, moving_image, weights]
+            ]
+        )
+    )
```

Comment:
I remember we discussed this part of the code as being not very straightforward (especially after black went through it).
I tried to make it a bit mor readable. Tests are still validated, and I did a quick %%timeit on both implementation and they are of similar speed.

```suggestion
    pyramid = []
    for x in [reference_image, moving_image, weights]:
        if x is None:
            levels = [None] * (max_layer + 1)

        else:
            levels = reversed(list(
                pyramid_gaussian(
                    x,
                    max_layer=max_layer,
                    downscale=pyramid_downscale,
                    preserve_range=True,
                    channel_axis=channel_axis,)))

        pyramid.append(levels)

    pyramid = list(zip(*pyramid))
```

Reply:
I'd actually transform it further. The `zip(*` is just an awkward way to get it all into a single expression. Here's my attempt (assume functools and itertools are imported):

```suggestion
    pyramid_fn = functools.partial(
        pyramid_gaussian,
        max_layer=max_layer,
        downscale=pyramid_downscale,
        preserve_range=True,
        channel_axis=channel_axis,
    )
    reference_pyramid = pyramid_fn(reference_image)
    moving_pyramid = pyramid_fn(moving_image)
    if weights is None:
        weights_pyramid = itertools.repeat(None)
    else:
        weights_pyramid = pyramid_fn(weights)
    pyramid = list(zip(reference_pyramid, moving_pyramid, weights_pyramid))
```